pkg/sethandler/controller.go

```diff
@@ -113,35 +115,71 @@ func (setHandler *SetHandler) PodChanged(oldPod, newPod v1.Pod) {
 	if !shouldPodBeHandled(newPod) {
 		return
 	}
-	setHandler.adjustContainerSets(newPod)
+	if proceed, containersToBeSet := determineContainersToBeSet(oldPod, newPod); proceed {
+			setHandler.adjustContainerSets(newPod, containersToBeSet)
+	}
 }
 
 func shouldPodBeHandled(pod v1.Pod) bool {
 	setterNodeName := os.Getenv("NODE_NAME")
 	podNodeName := pod.Spec.NodeName
-
 	//Pod is not yet scheduled, or it wasn't scheduled to the Node of this specific CPUSetter instance
 	if setterNodeName == "" || podNodeName == "" || setterNodeName != podNodeName {
 		return false
 	}
-	//If the Pod is not running, its containers haven't been created yet - no cpuset cgroup is present to be adjusted by the CPUSetter
-	if pod.ObjectMeta.Annotations[setterAnnotationKey] != "" {
-		return false
-	}
-	if len(pod.Status.ContainerStatuses) == 0 {
-		return false
-	}
-	for _, containerStatus := range pod.Status.ContainerStatuses {
-		if containerStatus.ContainerID == "" {
-			return false
+	return true
+}
+
+func determineContainersToBeSet(oldPod, newPod v1.Pod)  (bool, []string){
+	workingContainers := []string{}
+	found := false
+	if newPod.ObjectMeta.Annotations[setterAnnotationKey] != "" && oldPod.ObjectMeta.Name != ""  {
+		if len(newPod.Status.ContainerStatuses) == 0 || len(oldPod.Status.ContainerStatuses) == 0 {
+			return false, []string{}
+		}
+		for _, newContainerStatus := range newPod.Status.ContainerStatuses {
+			found = false
+			for _, oldContainerStatus := range oldPod.Status.ContainerStatuses {
+				if oldContainerStatus.ContainerID == newContainerStatus.ContainerID {
+					found = true
+				}
+			}
+			if !found {
+				workingContainers = append(workingContainers, newContainerStatus.Name)
+			}
+		}
+	} else {
+		if len(newPod.Status.ContainerStatuses) == 0 {
+			return false, []string{}
+		}
+		for _, newContainerStatus := range newPod.Status.ContainerStatuses {
+			if newContainerStatus.ContainerID == "" {
+				return false, []string{}
+			}
+			workingContainers = append(workingContainers, newContainerStatus.Name)
 		}
 	}
-	return true
+	if len(workingContainers) > 0 {
+			return true, workingContainers
+	}
+	return false, []string{}
 }
 
-func (setHandler *SetHandler) adjustContainerSets(pod v1.Pod) {
-	var pathToContainerCpusetFile string
+func (setHandler *SetHandler) adjustContainerSets(pod v1.Pod, containersToBeSet []string) {
+	var (
+		pathToContainerCpusetFile string
+		found bool
+	)
 	for _, container := range pod.Spec.Containers {
+		found = false
+		for _, checkContainer := range containersToBeSet{
```

Comment:
instead of ranging over a set all the time, you could use a map datastructure instead with the ContainerIDs being keys, the fact if they are needed to be set or not being a bool value
checking whether a container should be set or not would be a simple key check like this: https://github.com/nokia/danm-utils/blob/master/pkg/netruleset/netruleset.go#L52

Reply:
always choose data structures with the most frequent operation you want to execute on it kept in mind